---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

##  **qgg**
### An R package for Quantitative Genetic and Genomic analyses

**qgg** is an R package providing an environment for large-scale genetic analyses of quantitative and disease phenotypes.

**qgg** provides an infrastructure for efficient processing of large-scale genetic and phenotypic data including core functions for constructing marker-based genomic relationship matrices, estimating genetic parameters, performing genomic prediction and genetic risk profiling, and single marker and marker-set based association analyses.


The qgg package was developed based on the hypothesis that certain regions on the genome, so-called *genomic features*, may be enriched for causal variants affecting the trait. Several genomic feature classes can be formed based on previous studies and different sources of information including genes, chromosomes, biological pathways, gene ontologies, sequence annotation, prior QTL regions, or other types of external evidence.

The qgg package provides a range of genomic feature (linear mixed) modeling approaches for predicting quantitative trait phenotypes from high resolution genomic polymorphism data.

The models are implemented using likelihood or Bayesian methods. 

Genomic feature best linear unbiased prediction (GFBLUP) models can be fitted. We have extended these models to include multiple features and multiple traits. Different genetic models (e.g. additive, dominance, gene by gene and gene by environment interactions) can be used. Further extensions include a weighted GFBLUP model using differential weighting of the individual genetic marker relationships. 

Bayesian multiple trait and multiple genomic feature models can be fitted. The models are implemented using an empirical Bayesian method that handles multiple features and multiple traits. The models were implemented using spectral decomposition that plays an important computational role in the Markov chain Monte Carlo strategy. This is a very flexible and formal statistical framework for using prior information to decompose genomic (co)variances and predict trait phenotypes. 

The premise of the genomic feature models presented above is that genomic features are enriched for causal variants affecting the traits. However, in reality, the number, location and effect sizes of the true causal variants in the genomic feature are unknown. Therefore we have developed and evaluated a number of marker set tests. These approaches are computationally very fast allowing us to rapidly analyze different layers of genomic feature classes to discover genomic features potentially enriched for causal variants.  Results from these analyses are built into the above mentioned prediction models. 

