---
title: "Annotation of markers in the genomic relationship matrix"
author: "Izel Fourie SÃ¸rensen"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

knitr::opts_knit$set(root.dir = "C:/Users/Izel/Dropbox/qgg-usersguide")
```

Several critia can be used for defining marker sets, e.g., the gene ontology (GO) annotation of markers or the genes or chromosomes within which SNPs/markers occur. For this small data set tutorial we define marker sets based on the GO annotation of the markers. The GO of a gene is determined by the  function of a particular gene. Each GO annotation consists of an association between a gene and a GO term.  
In this script we prepare a data frame of the *Drosophila melanogaster* GO annotations.

#### Download and read the annotation data

Variant annotation (based on FB5.49) can be found at  <http://dgrp2.gnets.ncsu.edu/data.html> at the bottom of the page under "Other useful files". 

```{r, eval=FALSE}
download.file("http://dgrp2.gnets.ncsu.edu/data/website/dgrp.fb549.annot.txt", 
              destfile = "./data/dgrp.fb549.annot.txt")
```
  
```{r}
annotation <- read.table(file = "./data/dgrp.fb549.annot.txt", sep="\t",  
                         colClasses="character", quote="")
```
The file is tab separated. The separators used in the site class column (column 3) are explained here: 
<http://dgrp2.gnets.ncsu.edu/faq.html> under item "3. What are the output files?; GeneAnnotation".

  
Look at the data
```{r}
dim(annotation)
str(annotation)
```



#### Edit the annotation data frame

The first column of the `annotation` data frame is used as row names. `as.character()` is used to avoid factor levels.
```{r}
rownames(annotation) <- as.character(annotation[,1])
```

With `sapply` split each of the row names (e.g. 2L_10000016_SNP) at the underscore "_". Resulting in e.g. 2L 10000016 SNP.  The resulting third element of the row name (`[3]`, here "SNP") is saved in a vector called `vtype` (variant type).  
```{r}
vtype <- sapply(rownames(annotation), function(x){
  strsplit(x,split="_")[[1]][3]
  } )
head(vtype)
```
\hfill\break  
Rownames of the `annotation` data frame (SNP ids) are kept as names for each element in the `vtype` vector. See:
```{r}
head(names(vtype))
```
\hfill\break 
Look at unique terms in `vtype`.
```{r}
unique(vtype)
```
\hfill\break  
`snpA_raw` is a vector containing the information from column 3 (site class) of data frame `annotation`, including only SNPs. Look at the first element of the `snpA_raw` vector.
```{r}
snpA_raw <- as.character(annotation[vtype=="SNP",3])
length(snpA_raw)
snpA_raw[1]
```
\hfill\break 
Give names to the `snpA_raw` vector: the row names of the `annotation` data frame, i.e. SNP id. Remove the redundant suffix "_SNP" from the names of the vector, such that it only contains the SNP id.
```{r}
names(snpA_raw) <- gsub("_SNP","",rownames(annotation)[vtype=="SNP"])
```
\hfill\break  
Only the "SiteClass" information (i.e. information in the square brackets directly following "SiteClass") is used. This includes flybase gene id, gene symbol, mapped sequence ontology (site class) and base pair distance to gene and is separated by "," from the transcript annotation information (see: <http://dgrp2.gnets.ncsu.edu/faq.html>).  

Select the site class information by splitting `snpA_raw` at "," and keep the first string. Remove "SiteClass" and square brackets from this string and split the string at the semicolons.
```{r}
snpA_raw <- lapply(snpA_raw, function(x) {      
 x <- strsplit(x,",")[[1]][1]
 x <- gsub("SiteClass","",x)
 x <- gsub("[","",x, fixed=TRUE)
 x <- gsub("]","",x, fixed=TRUE)
 x <- strsplit(x,";")[[1]]
 x
  })

head(snpA_raw)
length(snpA_raw)
```
\hfill\break 
Create a vector (`nA_raw`) which contains the length of each of the SNPs (the number of separated segments per SNP) in `snpA_raw`.
```{r}
nA_raw <- sapply(snpA_raw,length)
```
  
`table(nA_raw)` shows the number of SNP loci that contain 1, 2, 3 or up to 12 separate annotations in flybase.
```{r}
table(nA_raw)
```
\hfill\break  
Look at the first three of 18 elements in `nA_raw` that has a length of e.g. 12.
```{r}
nA_raw[nA_raw==12][1:3]
```
\hfill\break  
The vector `snpNames` contains the names of snpA_raw repeated as many times as they appeared in nA_raw, i.e. once for each flybase annotation.
```{r}
snpNames <- rep(names(snpA_raw), times=nA_raw)
length(snpNames)
```

Unlist snpA_raw and remove the names of the vector so that the vector only contains the separated segments.  Unlisting `snpA_raw` creates a vector with the same length as `nA_raw`.
```{r}
snpA_raw <- unlist(snpA_raw, use.names=FALSE)
head(snpA_raw)
length(snpA_raw)
```

Split the segments of snpA_raw at the "|" symbol. The result is a list where the element names correspond to the separated segments prepared above. "GBgn" (flybase gene id), gene id, mapped sequence ontology terms ("SeqOnt") and base pair distance to gene ("distance") is the contents of each element. As earlier the vector `nA_raw` gives the length of each of the elements (the number of separated segments for each element) in `snpA_raw`. `table(nA_raw)` shows that 919806 elements in `snpA_raw` contain 3 segments and 3913325 elements contain 4 segments.
```{r}
snpA_raw <- sapply(snpA_raw,function(x) {  unlist(strsplit(x,split="|",fixed=TRUE))  })
nA_raw <- sapply(snpA_raw,length)
table(nA_raw)
```
  
Create an empty matrix `snpA`, the number of rows equal to the length of snpNames and 4 columns. The `snpNames` vector is used as the row names. This corresponds to element names of the `snpA_raw` list as prepared earlier.
```{r}
snpA <- matrix(NA,nrow=length(snpNames),ncol=4)
rownames(snpA) <- snpNames
head(snpA)
```
\hfill\break 
Unlist `snpA_raw`. This creates a vector of "FBgn", gene names, mapped sequence ontology terms ("SeqOnt") and base pair distance to gene ("distance"). By transposing the vector, the four different types of information are each put into a column of snpA.  Only the elements of snpA_raw that contain 4 segments (thus the elements of `nA_raw` with length 4) are included.
```{r}
snpA[nA_raw==4,] <- t(sapply(snpA_raw[nA_raw==4],function(x) {  unlist(x) }))
head(snpA)
```
\hfill\break 
The final annotation data frame will only include SNPs that are in the centered and scaled genotype matrix (**W**), as prepared earlier in the qgg user guide. Here we load the `W` matrix.

```{r}
load(file="./genotypes/dgrp2_W2.Rdata")
```
  
Prepare a logical vector, `inW` . `TRUE` depends on whether the SNPs in this dataset is also found in `W`.
```{r}
inW <- rownames(snpA)%in%colnames(W)
```
  
Keep only SNPs that were `TRUE` in the logical vector `inW`.
```{r}
snpA <- snpA[inW,]
```
  
Give the data frame snpA relevant column names.
```{r}
colnames(snpA) <- c("FBid","GeneName","SeqOnt","distance")
snpA[1:5,]
```
\hfill\break 
Save the snpA matrix
```{r}
save(snpA, file="./annotation/snpA_W2.Rdata")
```



