---
title: "Using qgg"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using qgg}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


## Installation

You can install qgg from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
library(devtools)
options(devtools.install.args=" --no-multiargs")
devtools::install_github("psoerensen/qgg")
```


## Small data set tutorial

This is a very detailed tutorial taking you all the way from getting and cleaning the data to linear mixed model analysis. It shows how to use **qgg** for the construction of a genomic relationship matrix, genomic best linear unbiased prediction (GBLUP), genomic feature BLUP analyses, and set tests.  
  
The analyses are performed on publicly available data on the [Drosophila genetic reference panel](http://dgrp2.gnets.ncsu.edu/) (DGRP).  Here are links to the different steps used for the analyses of this small scale data set.  

* [Prepare phenotype and covariate data](http://psoerensen.github.io/qgg/articles/Prepare_Phenotype_Data_Starvation.html)  
  
* [Compute a centered and scaled genotype matrix](http://psoerensen.github.io/qgg/articles/Compute_W.html) 
  
* [Gene ontology annotation of markers](http://psoerensen.github.io/qgg/articles/Prepare_Annotation.html)  
  
* [Prediction with GBLUP and cross validation](http://psoerensen.github.io/qgg/articles/GBLUP_cross_validation.html)  

  
