---
title: "Using qgg"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using qgg}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


## Installation

You can install **qgg** from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
library(devtools)
options(devtools.install.args=" --no-multiargs")
devtools::install_github("psoerensen/qgg")
```


## Small data set tutorial

This detailed tutorial on a small data set, exemplifies some of the core features of **qgg**. It takes you all the way from getting and cleaning data to estimation of genomic parameters, gene set enrichment analysis and making predictions with genomic best linear unbiased prediction (GBLUP) and genomic feature BLUP (GFBLUP) models.  
  
The phenotypic and genetic data is from the publicly available [*Drosophila* genetic reference panel](http://dgrp2.gnets.ncsu.edu/) (DGRP) - a system of 205 completely inbred *D. melanogaster* lines.  
  
The tutorial is divided in the following steps:
  
* **Data preparation**  
  
     + [Prepare phenotype and covariate data](http://psoerensen.github.io/qgg/articles/Prepare_Phenotype_Data_Starvation.html)  
     
     + [Prepare genotype data](http://psoerensen.github.io/qgg/articles/Prepare_Genotypes.html)  
     
     + [Prepare marker annotation](http://psoerensen.github.io/qgg/articles/Prepare_Annotation.html)  
  
     + [Prepare marker sets](http://psoerensen.github.io/qgg/articles/Prepare_Marker_Sets.html)
  
* **Data analysis**  

     + [GBLUP and cross validation](http://psoerensen.github.io/qgg/articles/GBLUP_cross_validation.html)  
  
     + [GFBLUP and cross validation](http://psoerensen.github.io/qgg/articles/GFBLUP_cross_validation.html)  
  
