<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotation of markers in the genomic relationship matrix • qgg</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Annotation of markers in the genomic relationship matrix">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qgg</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/qgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Small_data_set_tutorial.html">Small data set tutorial</a>
    </li>
    <li>
      <a href="../articles/Prepare_Phenotype_Data_Starvation.html">Prepare Phenotype and Covariate Data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="https://github.com/psoerensen/qgg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Annotation of markers in the genomic relationship matrix</h1>
                        <h4 class="author">Izel Fourie Sørensen</h4>
            
            <h4 class="date">02 November, 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/psoerensen/qgg/blob/master/vignettes/Prepare_Annotation.Rmd"><code>vignettes/Prepare_Annotation.Rmd</code></a></small>
      <div class="hidden name"><code>Prepare_Annotation.Rmd</code></div>

    </div>

    
    
<p>Several critia can be used for defining marker sets, e.g., the gene ontology (GO) annotation of markers or the genes or chromosomes within which SNPs/markers occur. For this small data set tutorial we define marker sets based on the GO annotation of the markers. The GO of a gene is determined by the function of a particular gene. Each GO annotation consists of an association between a gene and a GO term.<br>
In this script we prepare a data frame of the <em>Drosophila melanogaster</em> GO annotations.</p>
<div id="download-and-read-the-annotation-data" class="section level4">
<h4 class="hasAnchor">
<a href="#download-and-read-the-annotation-data" class="anchor"></a>Download and read the annotation data</h4>
<p>Variant annotation (based on FB5.49) can be found at <a href="http://dgrp2.gnets.ncsu.edu/data.html" class="uri">http://dgrp2.gnets.ncsu.edu/data.html</a> at the bottom of the page under “Other useful files”.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">download.file</span>(<span class="st">"http://dgrp2.gnets.ncsu.edu/data/website/dgrp.fb549.annot.txt"</span>, 
              <span class="dt">destfile =</span> <span class="st">"./data/dgrp.fb549.annot.txt"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">annotation &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="dt">file =</span> <span class="st">"./data/dgrp.fb549.annot.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,  
                         <span class="dt">colClasses=</span><span class="st">"character"</span>, <span class="dt">quote=</span><span class="st">""</span>)</code></pre></div>
<p>The file is tab separated. The separators used in the site class column (column 3) are explained here: <a href="http://dgrp2.gnets.ncsu.edu/faq.html" class="uri">http://dgrp2.gnets.ncsu.edu/faq.html</a> under item “3. What are the output files?; GeneAnnotation”.</p>
<p>Look at the data</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(annotation)</code></pre></div>
<pre><code>## [1] 4438427       4</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(annotation)</code></pre></div>
<pre><code>## 'data.frame':    4438427 obs. of  4 variables:
##  $ V1: chr  "2L_10000016_SNP" "2L_10000023_SNP" "2L_10000029_SNP" "2L_10000033_SNP" ...
##  $ V2: chr  "C" "C" "G" "G" ...
##  $ V3: chr  "SiteClass[FBgn0051875|CG31875|INTRON|0;FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0],TranscriptAnnot[INTRON(MODIFIE"| __truncated__ "SiteClass[FBgn0051875|CG31875|INTRON|0;FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0],TranscriptAnnot[INTRON(MODIFIE"| __truncated__ "SiteClass[FBgn0051875|CG31875|INTRON|0;FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0],TranscriptAnnot[INTRON(MODIFIE"| __truncated__ "SiteClass[FBgn0051875|CG31875|INTRON|0;FBgn0051755|SoYb|SYNONYMOUS_CODING|0],TranscriptAnnot[INTRON(MODIFIER|||"| __truncated__ ...
##  $ V4: chr  "-" "-" "-" "-" ...</code></pre>
</div>
<div id="edit-the-annotation-data-frame" class="section level4">
<h4 class="hasAnchor">
<a href="#edit-the-annotation-data-frame" class="anchor"></a>Edit the annotation data frame</h4>
<p>The first column of the <code>annotation</code> data frame is used as row names. <code>as.character()</code> is used to avoid factor levels.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rownames</span>(annotation) &lt;-<span class="st"> </span><span class="kw">as.character</span>(annotation[,<span class="dv">1</span>])</code></pre></div>
<p>With <code>sapply</code> split each of the row names (e.g. 2L_10000016_SNP) at the underscore “_“. Resulting in e.g. 2L 10000016 SNP. The resulting third element of the row name (<code>[3]</code>, here”SNP“) is saved in a vector called <code>vtype</code> (variant type).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vtype &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">rownames</span>(annotation), <span class="cf">function</span>(x){
  <span class="kw">strsplit</span>(x,<span class="dt">split=</span><span class="st">"_"</span>)[[<span class="dv">1</span>]][<span class="dv">3</span>]
  } )
<span class="kw">head</span>(vtype)</code></pre></div>
<pre><code>## 2L_10000016_SNP 2L_10000023_SNP 2L_10000029_SNP 2L_10000033_SNP 
##           "SNP"           "SNP"           "SNP"           "SNP" 
## 2L_10000089_SNP 2L_10000133_SNP 
##           "SNP"           "SNP"</code></pre>
<p> Rownames of the <code>annotation</code> data frame (SNP ids) are kept as names for each element in the <code>vtype</code> vector. See:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">names</span>(vtype))</code></pre></div>
<pre><code>## [1] "2L_10000016_SNP" "2L_10000023_SNP" "2L_10000029_SNP" "2L_10000033_SNP"
## [5] "2L_10000089_SNP" "2L_10000133_SNP"</code></pre>
<p> Look at unique terms in <code>vtype</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">unique</span>(vtype)</code></pre></div>
<pre><code>## [1] "SNP" "DEL" "INS" "MNP"</code></pre>
<p> <code>snpA_raw</code> is a vector containing the information from column 3 (site class) of data frame <code>annotation</code>, including only SNPs. Look at the first element of the <code>snpA_raw</code> vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA_raw &lt;-<span class="st"> </span><span class="kw">as.character</span>(annotation[vtype<span class="op">==</span><span class="st">"SNP"</span>,<span class="dv">3</span>])
<span class="kw">length</span>(snpA_raw)</code></pre></div>
<pre><code>## [1] 3963420</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA_raw[<span class="dv">1</span>]</code></pre></div>
<pre><code>## [1] "SiteClass[FBgn0051875|CG31875|INTRON|0;FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0],TranscriptAnnot[INTRON(MODIFIER||||161|CG31875|protein_coding|CODING|FBtr0303883|3);INTRON(MODIFIER||||161|CG31875|protein_coding|CODING|FBtr0303884|3);INTRON(MODIFIER||||161|CG31875|protein_coding|CODING|FBtr0308121|3);NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gaG/gaT|E471D|1476|SoYb|protein_coding|CODING|FBtr0303882|);NON_SYNONYMOUS_CODING(MODERATE|MISSENSE|gaG/gaT|E471D|1476|SoYb|protein_coding|CODING|FBtr0308122|)]"</code></pre>
<p> Give names to the <code>snpA_raw</code> vector: the row names of the <code>annotation</code> data frame, i.e. SNP id. Remove the redundant suffix “_SNP" from the names of the vector, such that it only contains the SNP id.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(snpA_raw) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"_SNP"</span>,<span class="st">""</span>,<span class="kw">rownames</span>(annotation)[vtype<span class="op">==</span><span class="st">"SNP"</span>])</code></pre></div>
<p> Only the “SiteClass” information (i.e. information in the square brackets directly following “SiteClass”) is used. This includes flybase gene id, gene symbol, mapped sequence ontology (site class) and base pair distance to gene and is separated by “,” from the transcript annotation information (see: <a href="http://dgrp2.gnets.ncsu.edu/faq.html" class="uri">http://dgrp2.gnets.ncsu.edu/faq.html</a>).</p>
<p>Select the site class information by splitting <code>snpA_raw</code> at “,” and keep the first string. Remove “SiteClass” and square brackets from this string and split the string at the semicolons.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA_raw &lt;-<span class="st"> </span><span class="kw">lapply</span>(snpA_raw, <span class="cf">function</span>(x) {      
 x &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x,<span class="st">","</span>)[[<span class="dv">1</span>]][<span class="dv">1</span>]
 x &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"SiteClass"</span>,<span class="st">""</span>,x)
 x &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"["</span>,<span class="st">""</span>,x, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
 x &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"]"</span>,<span class="st">""</span>,x, <span class="dt">fixed=</span><span class="ot">TRUE</span>)
 x &lt;-<span class="st"> </span><span class="kw">strsplit</span>(x,<span class="st">";"</span>)[[<span class="dv">1</span>]]
 x
  })

<span class="kw">head</span>(snpA_raw)</code></pre></div>
<pre><code>## $`2L_10000016`
## [1] "FBgn0051875|CG31875|INTRON|0"            
## [2] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## 
## $`2L_10000023`
## [1] "FBgn0051875|CG31875|INTRON|0"            
## [2] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## 
## $`2L_10000029`
## [1] "FBgn0051875|CG31875|INTRON|0"            
## [2] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## 
## $`2L_10000033`
## [1] "FBgn0051875|CG31875|INTRON|0"        
## [2] "FBgn0051755|SoYb|SYNONYMOUS_CODING|0"
## 
## $`2L_10000089`
## [1] "FBgn0051875|CG31875|INTRON|0"            
## [2] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## 
## $`2L_10000133`
## [1] "FBgn0051875|CG31875|INTRON|0"        
## [2] "FBgn0051755|SoYb|SYNONYMOUS_CODING|0"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(snpA_raw)</code></pre></div>
<pre><code>## [1] 3963420</code></pre>
<p> Create a vector (<code>nA_raw</code>) which contains the length of each of the SNPs (the number of separated segments per SNP) in <code>snpA_raw</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nA_raw &lt;-<span class="st"> </span><span class="kw">sapply</span>(snpA_raw,length)</code></pre></div>
<p><code>table(nA_raw)</code> shows the number of SNP loci that contain 1, 2, 3 or up to 12 separate annotations in flybase.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(nA_raw)</code></pre></div>
<pre><code>## nA_raw
##       1       2       3       4       5       6       7       8       9 
## 3241772  602432  100328   13818    2828     963     494     440     170 
##      10      11      12 
##     130      27      18</code></pre>
<p> Look at the first three of 18 elements in <code>nA_raw</code> that has a length of e.g. 12.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nA_raw[nA_raw<span class="op">==</span><span class="dv">12</span>][<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</code></pre></div>
<pre><code>## 2L_20418974 2L_20418976 2L_20418995 
##          12          12          12</code></pre>
<p> The vector <code>snpNames</code> contains the names of snpA_raw repeated as many times as they appeared in nA_raw, i.e. once for each flybase annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpNames &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="kw">names</span>(snpA_raw), <span class="dt">times=</span>nA_raw)
<span class="kw">length</span>(snpNames)</code></pre></div>
<pre><code>## [1] 4833131</code></pre>
<p>Unlist snpA_raw and remove the names of the vector so that the vector only contains the separated segments. Unlisting <code>snpA_raw</code> creates a vector with the same length as <code>nA_raw</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA_raw &lt;-<span class="st"> </span><span class="kw">unlist</span>(snpA_raw, <span class="dt">use.names=</span><span class="ot">FALSE</span>)
<span class="kw">head</span>(snpA_raw)</code></pre></div>
<pre><code>## [1] "FBgn0051875|CG31875|INTRON|0"            
## [2] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## [3] "FBgn0051875|CG31875|INTRON|0"            
## [4] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"
## [5] "FBgn0051875|CG31875|INTRON|0"            
## [6] "FBgn0051755|SoYb|NON_SYNONYMOUS_CODING|0"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(snpA_raw)</code></pre></div>
<pre><code>## [1] 4833131</code></pre>
<p>Split the segments of snpA_raw at the “|” symbol. The result is a list where the element names correspond to the separated segments prepared above. “GBgn” (flybase gene id), gene id, mapped sequence ontology terms (“SeqOnt”) and base pair distance to gene (“distance”) is the contents of each element. As earlier the vector <code>nA_raw</code> gives the length of each of the elements (the number of separated segments for each element) in <code>snpA_raw</code>. <code>table(nA_raw)</code> shows that 919806 elements in <code>snpA_raw</code> contain 3 segments and 3913325 elements contain 4 segments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA_raw &lt;-<span class="st"> </span><span class="kw">sapply</span>(snpA_raw,<span class="cf">function</span>(x) {  <span class="kw">unlist</span>(<span class="kw">strsplit</span>(x,<span class="dt">split=</span><span class="st">"|"</span>,<span class="dt">fixed=</span><span class="ot">TRUE</span>))  })
nA_raw &lt;-<span class="st"> </span><span class="kw">sapply</span>(snpA_raw,length)
<span class="kw">table</span>(nA_raw)</code></pre></div>
<pre><code>## nA_raw
##       3       4 
##  919806 3913325</code></pre>
<p>Create an empty matrix <code>snpA</code>, the number of rows equal to the length of snpNames and 4 columns. The <code>snpNames</code> vector is used as the row names. This corresponds to element names of the <code>snpA_raw</code> list as prepared earlier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>,<span class="dt">nrow=</span><span class="kw">length</span>(snpNames),<span class="dt">ncol=</span><span class="dv">4</span>)
<span class="kw">rownames</span>(snpA) &lt;-<span class="st"> </span>snpNames
<span class="kw">head</span>(snpA)</code></pre></div>
<pre><code>##             [,1] [,2] [,3] [,4]
## 2L_10000016   NA   NA   NA   NA
## 2L_10000016   NA   NA   NA   NA
## 2L_10000023   NA   NA   NA   NA
## 2L_10000023   NA   NA   NA   NA
## 2L_10000029   NA   NA   NA   NA
## 2L_10000029   NA   NA   NA   NA</code></pre>
<p> Unlist <code>snpA_raw</code>. This creates a vector of “FBgn”, gene names, mapped sequence ontology terms (“SeqOnt”) and base pair distance to gene (“distance”). By transposing the vector, the four different types of information are each put into a column of snpA. Only the elements of snpA_raw that contain 4 segments (thus the elements of <code>nA_raw</code> with length 4) are included.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA[nA_raw<span class="op">==</span><span class="dv">4</span>,] &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">sapply</span>(snpA_raw[nA_raw<span class="op">==</span><span class="dv">4</span>],<span class="cf">function</span>(x) {  <span class="kw">unlist</span>(x) }))
<span class="kw">head</span>(snpA)</code></pre></div>
<pre><code>##             [,1]          [,2]      [,3]                    [,4]
## 2L_10000016 "FBgn0051875" "CG31875" "INTRON"                "0" 
## 2L_10000016 "FBgn0051755" "SoYb"    "NON_SYNONYMOUS_CODING" "0" 
## 2L_10000023 "FBgn0051875" "CG31875" "INTRON"                "0" 
## 2L_10000023 "FBgn0051755" "SoYb"    "NON_SYNONYMOUS_CODING" "0" 
## 2L_10000029 "FBgn0051875" "CG31875" "INTRON"                "0" 
## 2L_10000029 "FBgn0051755" "SoYb"    "NON_SYNONYMOUS_CODING" "0"</code></pre>
<p> The final annotation data frame will only include SNPs that are in the centered and scaled genotype matrix (<strong>W</strong>), as prepared earlier in the qgg user guide. Here we load the <code>W</code> matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="dt">file=</span><span class="st">"./genotypes/dgrp2_W2.Rdata"</span>)</code></pre></div>
<p>Prepare a logical vector, <code>inW</code> . <code>TRUE</code> depends on whether the SNPs in this dataset is also found in <code>W</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inW &lt;-<span class="st"> </span><span class="kw">rownames</span>(snpA)<span class="op">%in%</span><span class="kw">colnames</span>(W)</code></pre></div>
<p>Keep only SNPs that were <code>TRUE</code> in the logical vector <code>inW</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">snpA &lt;-<span class="st"> </span>snpA[inW,]</code></pre></div>
<p>Give the data frame snpA relevant column names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(snpA) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"FBid"</span>,<span class="st">"GeneName"</span>,<span class="st">"SeqOnt"</span>,<span class="st">"distance"</span>)
snpA[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,]</code></pre></div>
<pre><code>##             FBid          GeneName  SeqOnt                  distance
## 2L_10000016 "FBgn0051875" "CG31875" "INTRON"                "0"     
## 2L_10000016 "FBgn0051755" "SoYb"    "NON_SYNONYMOUS_CODING" "0"     
## 2L_10000033 "FBgn0051875" "CG31875" "INTRON"                "0"     
## 2L_10000033 "FBgn0051755" "SoYb"    "SYNONYMOUS_CODING"     "0"     
## 2L_10000089 "FBgn0051875" "CG31875" "INTRON"                "0"</code></pre>
<p> Save the snpA matrix</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(snpA, <span class="dt">file=</span><span class="st">"./annotation/snpA_W2.Rdata"</span>)</code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
