<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Prepare genotype data for all statistical analyses (initial step) — gprep • qgg</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Prepare genotype data for all statistical analyses (initial step) — gprep" />

<meta property="og:description" content="All functions in qgg relies on a simple data infrastructure that takes five main input sources; 
phenotype data (y), covariate data (X), genotype data (G or Glist), a genomic relationship 
matrix (GRM or GRMlist) and genetic marker sets (sets).
The genotypes are stored in a matrix (n x m (individuals x markers)) in memory (G) or in a 
binary file on disk (Glist).
It is only for small data sets that the genotype matrix (G) can stored in memory. For large data
sets the genotype matrix has to stored in a binary file on disk (Glist). Glist is as a list 
structure that contains information about the genotypes in the binary file.
The gprep function prepares the Glist, and is required for downstream analyses of large-scale 
genetic data. Typically, the Glist is prepared once, and saved as an *.Rdata-file.
The gprep function reads genotype information from binary PLINK files, and creates the Glist 
object that contains general information about the genotypes such as reference alleles, 
allele frequencies and missing genotypes, and construct a binary file on the disk that contains 
the genotypes as allele counts of the alternative allele (memory usage = (n x m)/4 bytes).
The Glist structure is used as input parameter for a number of qgg core functions including: 
1) construction of genomic relationship matrices (grm), 2) estimating genomic parameters (greml), 
3) single marker association analyses (lma or mlma), 4) gene set enrichment analyses (gsea), 
and 5) genomic prediction from genotypes and phenotypes (gsolve) or genotypes and summary statistics (gscore)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qgg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/qgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="https://github.com/psoerensen/qgg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Prepare genotype data for all statistical analyses (initial step)</h1>
    <small class="dont-index">Source: <a href='https://github.com/psoerensen/qgg/blob/master/R/compute_genotype_matrix.R'><code>R/compute_genotype_matrix.R</code></a></small>
    <div class="hidden name"><code>gprep.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>All functions in qgg relies on a simple data infrastructure that takes five main input sources; 
phenotype data (y), covariate data (X), genotype data (G or Glist), a genomic relationship 
matrix (GRM or GRMlist) and genetic marker sets (sets).</p>
<p>The genotypes are stored in a matrix (n x m (individuals x markers)) in memory (G) or in a 
binary file on disk (Glist).</p>
<p>It is only for small data sets that the genotype matrix (G) can stored in memory. For large data
sets the genotype matrix has to stored in a binary file on disk (Glist). Glist is as a list 
structure that contains information about the genotypes in the binary file.</p>
<p>The gprep function prepares the Glist, and is required for downstream analyses of large-scale 
genetic data. Typically, the Glist is prepared once, and saved as an *.Rdata-file.</p>
<p>The gprep function reads genotype information from binary PLINK files, and creates the Glist 
object that contains general information about the genotypes such as reference alleles, 
allele frequencies and missing genotypes, and construct a binary file on the disk that contains 
the genotypes as allele counts of the alternative allele (memory usage = (n x m)/4 bytes).</p>
<p>The Glist structure is used as input parameter for a number of qgg core functions including: 
1) construction of genomic relationship matrices (grm), 2) estimating genomic parameters (greml), 
3) single marker association analyses (lma or mlma), 4) gene set enrichment analyses (gsea), 
and 5) genomic prediction from genotypes and phenotypes (gsolve) or genotypes and summary statistics (gscore).</p>
    
    </div>

    <pre class="usage"><span class='fu'>gprep</span>(<span class='kw'>study</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>fnRAW</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bedfiles</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>bimfiles</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>famfiles</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ids</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rsids</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>overwrite</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>study</th>
      <td><p>name of the study</p></td>
    </tr>
    <tr>
      <th>fnRAW</th>
      <td><p>path and .raw filename of the binary file used for storing genotypes on the disk</p></td>
    </tr>
    <tr>
      <th>bedfiles</th>
      <td><p>vector of names for the PLINK bed-files</p></td>
    </tr>
    <tr>
      <th>bimfiles</th>
      <td><p>vector of names for the PLINK bim-files</p></td>
    </tr>
    <tr>
      <th>famfiles</th>
      <td><p>vector of names for the PLINK fam-files</p></td>
    </tr>
    <tr>
      <th>ids</th>
      <td><p>vector of individuals used in the study</p></td>
    </tr>
    <tr>
      <th>rsids</th>
      <td><p>vector of marker rsids used in the study</p></td>
    </tr>
    <tr>
      <th>overwrite</th>
      <td><p>logical if TRUE overwite binary genotype file</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>number of cores used to process the genotypes</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a list structure with information about genotypes</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"># NOT RUN {
#Glist <- gprep( bedfiles, bimfiles, study, path, additional arguments...)
#W <- getW( Glist, ids, rsids, additional arguments...)
# }
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Peter Soerensen</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

