<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Gene set enrichment analysis — gsea • qgg</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Gene set enrichment analysis — gsea" />

<meta property="og:description" content="The function gsea can perform several different gene set enrichment analyses. The general procedure is to obtain 
single marker statistics (e.g. summary statistics), from which it is possible to compute and evaluate a test statistic 
for a set of genetic markers that measures a joint degree of association between the marker set and the phenotype. 
The marker set is defined by a genomic feature such as genes, biological pathways, gene interactions, 
gene expression profiles etc.
Currently, four types of gene set enrichment analyses can be conducted with gsea; sum-based, count-based, 
score-based, and our own developed method, the covariance association test (CVAT). For details and comparisons of 
test statistics consult doi:10.1534/genetics.116.189498.
The sum test is based on the sum of all marker summary statistics located within the feature set. The single marker 
summary statistics can be obtained from linear model analyses (from PLINK or using the qgg lma approximation), 
or from single or multiple component REML analyses (GBLUP or GFBLUP) from the greml function. The sum test is powerful 
if the genomic feature harbors many genetic markers that have small to moderate effects.
The count-based method is based on counting the number of markers within a genomic feature that show association 
(or have single marker p-value below a certain threshold) with the phenotype. Under the null hypothesis (that the 
associated markers are picked at random from the total number of markers, thus, no enrichment of markers in any 
genomic feature) it is assumed that the observed count statistic is a realization from a hypergeometric distribution.
The score-based approach is based on the product between the scaled genotypes in a genomic feature and the residuals 
from the liner mixed model (obtained from greml).
The covariance association test (CVAT) is derived from the fit object from greml (GBLUP or GFBLUP), and measures 
the covariance between the total genomic effects for all markers and the genomic effects of the markers within the 
genomic feature.
The distribution of the test statistics obtained from the sum-based, score-based and CVAT is unknown, therefore 
a circular permutation approach is used to obtain an empirical distribution of test statistics." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qgg</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/qgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Small_data_set_tutorial.html">Small data set tutorial</a>
    </li>
    <li>
      <a href="../articles/prepare_phenotype_covariate_data.html">Prepare phenotype and covariate data</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="https://github.com/psoerensen/qgg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Gene set enrichment analysis</h1>
    <small class="dont-index">Source: <a href='https://github.com/psoerensen/qgg/blob/master/R/multiple_marker_set_test.R'><code>R/multiple_marker_set_test.R</code></a></small>
    <div class="hidden name"><code>gsea.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>The function gsea can perform several different gene set enrichment analyses. The general procedure is to obtain 
single marker statistics (e.g. summary statistics), from which it is possible to compute and evaluate a test statistic 
for a set of genetic markers that measures a joint degree of association between the marker set and the phenotype. 
The marker set is defined by a genomic feature such as genes, biological pathways, gene interactions, 
gene expression profiles etc.</p>
<p>Currently, four types of gene set enrichment analyses can be conducted with gsea; sum-based, count-based, 
score-based, and our own developed method, the covariance association test (CVAT). For details and comparisons of 
test statistics consult doi:10.1534/genetics.116.189498.</p>
<p>The sum test is based on the sum of all marker summary statistics located within the feature set. The single marker 
summary statistics can be obtained from linear model analyses (from PLINK or using the qgg lma approximation), 
or from single or multiple component REML analyses (GBLUP or GFBLUP) from the greml function. The sum test is powerful 
if the genomic feature harbors many genetic markers that have small to moderate effects.</p>
<p>The count-based method is based on counting the number of markers within a genomic feature that show association 
(or have single marker p-value below a certain threshold) with the phenotype. Under the null hypothesis (that the 
associated markers are picked at random from the total number of markers, thus, no enrichment of markers in any 
genomic feature) it is assumed that the observed count statistic is a realization from a hypergeometric distribution.</p>
<p>The score-based approach is based on the product between the scaled genotypes in a genomic feature and the residuals 
from the liner mixed model (obtained from greml).</p>
<p>The covariance association test (CVAT) is derived from the fit object from greml (GBLUP or GFBLUP), and measures 
the covariance between the total genomic effects for all markers and the genomic effects of the markers within the 
genomic feature.</p>
<p>The distribution of the test statistics obtained from the sum-based, score-based and CVAT is unknown, therefore 
a circular permutation approach is used to obtain an empirical distribution of test statistics.</p>
    
    </div>

    <pre class="usage"><span class='fu'>gsea</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>,
  <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>1000</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>stat</th>
      <td><p>vector or matrix of single marker statistics (e.g. coefficients, t-statistics, p-values)</p></td>
    </tr>
    <tr>
      <th>sets</th>
      <td><p>list of marker sets - names corresponds to row names in stat</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>used if method='hyperg' (threshold=0.05 is default)</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>including sum, cvat, hyperg, score</p></td>
    </tr>
    <tr>
      <th>nperm</th>
      <td><p>number of permutations used for obtaining an empirical p-value</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>number of cores used in the analysis</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>matrix of centered and scaled genotypes (used if method = cvat or score)</p></td>
    </tr>
    <tr>
      <th>fit</th>
      <td><p>list object obtained from a linear mixed model fit using the greml function</p></td>
    </tr>
    <tr>
      <th>g</th>
      <td><p>vector (or matrix) of genetic effects obtained from a linear mixed model fit (GBLUP of GFBLUP)</p></td>
    </tr>
    <tr>
      <th>s</th>
      <td><p>vector (or list) of single marker effects obtained from a linear mixed model fit (GBLUP of GFBLUP)</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a dataframe or a list including</p>
<dt>stat</dt><dd><p>marker set test statistics</p></dd> 
<dt>m</dt><dd><p>number of markers in the set</p></dd>
<dt>p</dt><dd><p>enrichment p-value for marker set</p></dd>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>

 <span class='co'># Simulate data</span>
 <span class='no'>W</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='fl'>20000000</span>), <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>10000</span>)
 <span class='fu'>colnames</span>(<span class='no'>W</span>) <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fu'>ncol</span>(<span class='no'>W</span>))
 <span class='fu'>rownames</span>(<span class='no'>W</span>) <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fu'>nrow</span>(<span class='no'>W</span>))
 <span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1</span>:<span class='fl'>10</span>]) + <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1001</span>:<span class='fl'>1010</span>]) + <span class='fu'>rnorm</span>(<span class='fu'>nrow</span>(<span class='no'>W</span>))

 <span class='co'># Create model</span>
 <span class='no'>data</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>y</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>1</span>)
 <span class='no'>fm</span> <span class='kw'>&lt;-</span> <span class='no'>y</span> ~ <span class='fl'>0</span> + <span class='no'>mu</span>
 <span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>model.matrix</span>(<span class='no'>fm</span>, <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>)

 <span class='co'># Single marker association analyses</span>
 <span class='no'>ma</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='lma.html'>lma</a></span>(<span class='kw'>y</span><span class='kw'>=</span><span class='no'>y</span>,<span class='kw'>X</span><span class='kw'>=</span><span class='no'>X</span>,<span class='kw'>W</span><span class='kw'>=</span><span class='no'>W</span>)

 <span class='co'># Create marker sets</span>
 <span class='no'>f</span> <span class='kw'>&lt;-</span> <span class='fu'>factor</span>(<span class='fu'>rep</span>(<span class='fl'>1</span>:<span class='fl'>100</span>,<span class='kw'>each</span><span class='kw'>=</span><span class='fl'>100</span>), <span class='kw'>levels</span><span class='kw'>=</span><span class='fl'>1</span>:<span class='fl'>100</span>)
 <span class='no'>sets</span> <span class='kw'>&lt;-</span> <span class='fu'>split</span>(<span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fl'>10000</span>),<span class='kw'>f</span><span class='kw'>=</span><span class='no'>f</span>)

 <span class='co'># Set test based on sums </span>
 <span class='no'>mma</span> <span class='kw'>&lt;-</span> <span class='fu'>gsea</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>ma</span>[,<span class='st'>"stat"</span>]**<span class='fl'>2</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>10000</span>)
 <span class='fu'>head</span>(<span class='no'>mma</span>)

 <span class='co'># Set test based on hyperG </span>
 <span class='no'>mma</span> <span class='kw'>&lt;-</span> <span class='fu'>gsea</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>ma</span>[,<span class='st'>"p"</span>], <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"hyperg"</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)
 <span class='fu'>head</span>(<span class='no'>mma</span>)


<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Peter Soerensen

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Peter Soerensen, Palle Duun Rohde, Izel Fourie Soerensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

