<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Multi marker association analysis based on marker set tests — mma • qgg</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Multi marker association analysis based on marker set tests — mma" />

<meta property="og:description" content="Genetic marker set tests based on summing the single genetic marker test statistics.
The single marker test statistics can be obtained from GBLUP and GFBLUP model fits or from standard GWAS." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qgg</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/qgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Small_data_set_tutorial.html">Small data set tutorial</a>
    </li>
    <li>
      <a href="../articles/Prepare_Phenotype_Data_Starvation.html">Prepare Phenotype and Covariate Data</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="https://github.com/psoerensen/qgg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Multi marker association analysis based on marker set tests</h1>
    <small class="dont-index">Source: <a href='https://github.com/psoerensen/qgg/blob/master/R/multiple_marker_set_test.R'><code>R/multiple_marker_set_test.R</code></a></small>
    <div class="hidden name"><code>mma.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Genetic marker set tests based on summing the single genetic marker test statistics.
The single marker test statistics can be obtained from GBLUP and GFBLUP model fits or from standard GWAS.</p>
    
    </div>

    <pre class="usage"><span class='fu'>mma</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>stat</th>
      <td><p>vector of single marker statistics (e.g. marker effects, t-stat, p-values)</p></td>
    </tr>
    <tr>
      <th>sets</th>
      <td><p>list of marker sets - names corresponds to rownames in stat</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>number of cores</p></td>
    </tr>
    <tr>
      <th>nperm</th>
      <td><p>number of permutations</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>including sum, cvat, hyperG, score</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>matrix of centered and scaled genotypes (used if method = cvat or score)</p></td>
    </tr>
    <tr>
      <th>fit</th>
      <td><p>is the fit object obtained from a linear mixed model fit using the greml function</p></td>
    </tr>
    <tr>
      <th>g</th>
      <td><p>vector (or list) of genetic effects obtained from a linear mixed model fit (GBLUP of GFBLUP)</p></td>
    </tr>
    <tr>
      <th>s</th>
      <td><p>vector (or list) of single marker effects obtained from a linear mixed model fit (GBLUP of GFBLUP)</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>used if method = hyperG</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a dataframe including</p>
<dt>setT</dt><dd><p>marker set test statistics</p></dd> 
<dt>nset</dt><dd><p>number of markers in the set</p></dd>
<dt>p</dt><dd><p>p-value for marker set</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The sum test is powerful if the genomic feature harbors many genetic markers having small to moderate effects. 
The distribution of this test statistic under the null hypothesis (associated markers are picked at random from the total 
number of tested genetic markers) is difficult to describe in terms of exact or approximate 
distributions, and an empirical distribution is required.
Genetic marker set tests based on the covariance statistics for a set of genetic markers.
The covariance test statistic is derived from a GBLUP (or GFBLUP) model fit. It is a measure of covariance between the total genomic effect for all markers 
and the genomic effect for the genetic markers in the genomic feature. It also relates to the explained sums of
squares for the genetic markers. 
The distribution of these test statistics under the null hypothesis is difficult to describe in terms of exact or approximate 
distributions, and an empirical distribution is required.
Genetic marker set tests based on the hyperG test statistics for a set of genetic markers.
The hyperG marker set test tests a predefined set of markers (i.e. those within a particular genomic feature)
for an association with the trait phenotype.
Under the null hypothesis (associated markers are picked at random from the total number of tested 
genetic markers) it is assumed that the observed count statistic is a realization from a hypergeometric 
distribution.
This hypothesis can be formulated and tested in a number of ways. Here we consider a test statistic based 
on counting the number of genetic markers in the feature that are associated to trait phenotype. 
A test based on the count test statistic is likely to have high power to detect association if the genomic 
feature harbours genetic markers with large effects.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Rohde, P. D., Demontis, D., Cuyabano, B. C. D., Boerglum, A. D., &amp; Soerensen, P. (2016). Covariance Association Test (CVAT) Identifies Genetic Markers Associated with Schizophrenia in Functionally Associated Biological Processes. Genetics, 203(4), 1901-1913.</p>
<p>Rohde, P. D., Edwards S. M., Sarup P., Soerensen, P. (August, 2014). Gene-based Association Approach Identify Genes Across Stress Traits in Fruit Flies. Poster presented at the 10th World Congress of Genetics Applied to Livestock Production (WCGALP), Vancouver, Canada.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>

 <span class='co'># Simulate data</span>
 <span class='no'>W</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='fl'>20000000</span>), <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>10000</span>)
 <span class='fu'>colnames</span>(<span class='no'>W</span>) <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fu'>ncol</span>(<span class='no'>W</span>))
 <span class='fu'>rownames</span>(<span class='no'>W</span>) <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fu'>nrow</span>(<span class='no'>W</span>))
 <span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1</span>:<span class='fl'>10</span>]) + <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1001</span>:<span class='fl'>1010</span>]) + <span class='fu'>rnorm</span>(<span class='fu'>nrow</span>(<span class='no'>W</span>))

 <span class='co'># Create model</span>
 <span class='no'>data</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>y</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>1</span>)
 <span class='no'>fm</span> <span class='kw'>&lt;-</span> <span class='no'>y</span> ~ <span class='fl'>0</span> + <span class='no'>mu</span>
 <span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>model.matrix</span>(<span class='no'>fm</span>, <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>)

 <span class='co'># Compute GRM</span>
 <span class='no'>GRM</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='computeGRM.html'>computeGRM</a></span>(<span class='kw'>W</span> <span class='kw'>=</span> <span class='no'>W</span>)

 <span class='co'># REML analyses and single marker association test</span>
 <span class='no'>fit</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='greml.html'>greml</a></span>(<span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>y</span>, <span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>X</span>, <span class='kw'>GRM</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='no'>GRM</span>), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; [1] "Iteration:" "1"          "Theta:"     "2.25"       "2.01"      
#&gt; [1] "Iteration:" "2"          "Theta:"     "4.4"        "3.2"       
#&gt; [1] "Iteration:" "3"          "Theta:"     "8.2"        "4.12"      
#&gt; [1] "Iteration:" "4"          "Theta:"     "13.45"      "3.69"      
#&gt; [1] "Iteration:" "5"          "Theta:"     "17.53"      "2.36"      
#&gt; [1] "Iteration:" "6"          "Theta:"     "18.59"      "1.87"      
#&gt; [1] "Iteration:" "7"          "Theta:"     "18.64"      "1.84"      
#&gt; [1] "Iteration:" "8"          "Theta:"     "18.64"      "1.84"      
#&gt; [1] "Iteration:" "9"          "Theta:"     "18.64"      "1.84"      
#&gt; [1] "Iteration:" "10"         "Theta:"     "18.64"      "1.84"      
#&gt; [1] "Converged at Iteration:" "10"                     
#&gt; [3] "Theta:"                  "18.64"                  
#&gt; [5] "1.84"                   </div><div class='input'> <span class='no'>ma</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='mlma.html'>mlma</a></span>(<span class='kw'>fit</span> <span class='kw'>=</span> <span class='no'>fit</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='no'>W</span>)
 <span class='no'>ma</span> <span class='kw'>&lt;-</span> <span class='fu'>as.matrix</span>(<span class='no'>ma</span>)

 <span class='no'>f</span> <span class='kw'>&lt;-</span> <span class='fu'>factor</span>(<span class='fu'>rep</span>(<span class='fl'>1</span>:<span class='fl'>100</span>,<span class='kw'>each</span><span class='kw'>=</span><span class='fl'>100</span>), <span class='kw'>levels</span><span class='kw'>=</span><span class='fl'>1</span>:<span class='fl'>100</span>)
 <span class='no'>sets</span> <span class='kw'>&lt;-</span> <span class='fu'>split</span>(<span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fl'>10000</span>),<span class='kw'>f</span><span class='kw'>=</span><span class='no'>f</span>)

<span class='co'># Set test based on sums </span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>mma</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>ma</span>[,<span class='st'>"stat"</span>]**<span class='fl'>2</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>10000</span>)</div><div class='output co'>#&gt; <span class='error'>Error in .Fortran("psets", m = as.integer(m), stat = as.double(stat),     nsets = as.integer(nsets), setstat = as.double(setstat),     msets = as.integer(msets), p = as.integer(rep(0, nsets)),     np = as.integer(np), ncores = as.integer(ncores), PACKAGE = "qgg"): "psets" not available for .Fortran() for package "qgg"</span></div><div class='input'>

<span class='co'># Set test based on hyperG </span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>mma</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>ma</span>[,<span class='st'>"p"</span>], <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"hyperG"</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)</div><div class='output co'>#&gt; <span class='error'>Error in mma(stat = ma[, "p"], sets = sets, method = "hyperG", threshold = 0.05): unused argument (threshold = 0.05)</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Peter Soerensen

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Peter Soerensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

