<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genetic marker set tests based on sum statistics — setTest • qgg</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Genetic marker set tests based on sum statistics — setTest" />

<meta property="og:description" content="Set test based on summing the single genetic marker test statistics.
The sum test is powerful if the genomic feature harbors many genetic markers having small to moderate effects." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">qgg</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/qgg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Prepare_Phenotype_Data_Starvation_html.html">Prepare Phenotype and Covariate Data</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/psoerensen/qgg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genetic marker set tests based on sum statistics</h1>
    <small class="dont-index">Source: <a href='https://github.com/psoerensen/qgg/blob/master/R/multiple_marker_set_test.R'><code>R/multiple_marker_set_test.R</code></a></small>
    <div class="hidden name"><code>setTest.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Set test based on summing the single genetic marker test statistics.
The sum test is powerful if the genomic feature harbors many genetic markers having small to moderate effects.</p>
    
    </div>

    <pre class="usage"><span class='fu'>setTest</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>stat</th>
      <td><p>vector of single marker statistics (e.g. marker effects, t-stat, p-values)</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>matrix of centered and scaled genotypes (used if method = cvat or score)</p></td>
    </tr>
    <tr>
      <th>sets</th>
      <td><p>list of marker sets - names corresponds to rownames in stat</p></td>
    </tr>
    <tr>
      <th>nperm</th>
      <td><p>number of permutations</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>including sum, cvat, hyperG, score</p></td>
    </tr>
    <tr>
      <th>threshold</th>
      <td><p>used if method = hyperG</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns a dataframe including</p>
<dt>setT</dt><dd><p>marker set test statistics</p></dd> 
<dt>nset</dt><dd><p>number of markers in the set</p></dd>
<dt>p</dt><dd><p>p-value for marker set</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The singler marker test statistics can be obtained from GBLUP and GFBLUP model fits or from standard GWAS. 
The distribution of this test statistic under the null hypothesis (associated markers are picked at random from the total 
number of tested genetic markers) is difficult to describe in terms of exact or approximate 
distributions, and an empirical distribution is required.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># Simulate data</span>
<span class='no'>W</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='fl'>2000000</span>), <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>10000</span>)
  <span class='fu'>colnames</span>(<span class='no'>W</span>) <span class='kw'>&lt;-</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fu'>ncol</span>(<span class='no'>W</span>))
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1</span>:<span class='fl'>10</span>]) + <span class='fu'>rowSums</span>(<span class='no'>W</span>[, <span class='fl'>1001</span>:<span class='fl'>1010</span>]) + <span class='fu'>rnorm</span>(<span class='fu'>nrow</span>(<span class='no'>W</span>))

<span class='co'># REML analyses </span>
<span class='no'>data</span> <span class='kw'>&lt;-</span> <span class='fu'>data.frame</span>(<span class='kw'>y</span> <span class='kw'>=</span> <span class='no'>y</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='no'>fm</span> <span class='kw'>&lt;-</span> <span class='no'>y</span> ~ <span class='no'>mu</span>
<span class='no'>fit</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='gfm.html'>gfm</a></span>(<span class='kw'>fm</span> <span class='kw'>=</span> <span class='no'>fm</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='no'>W</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='fu'>colnames</span>(<span class='no'>W</span>)), <span class='kw'>data</span> <span class='kw'>=</span> <span class='no'>data</span>)</div><div class='output co'>#&gt; start algorithm at 11.8153 11.8153 
#&gt; 1 sigma = 11.81529 11.81529 resid llik = -413.1626 
#&gt; 1 adjusted sigma = 11.7223 11.7223
#&gt; 2 sigma = 24.12272 5.73551 resid llik = -412.8265 
#&gt; 2 adjusted sigma = 18.95108 4.505881 delta.llik = 0.3360854 
#&gt; 3 sigma = 25.94281 1.227899 resid llik = -412.743 
#&gt; 3 adjusted sigma = 22.43548 1.061894 delta.llik = 0.08350128 
#&gt; 4 sigma = 26.87045 0.02522451 resid llik = -412.7278 
#&gt; 4 adjusted sigma = 23.49198 0.02205299 delta.llik = 0.01521621 
#&gt; 5 sigma = 26.64627 2.061154e-09 resid llik = -412.7275 
#&gt; 5 adjusted sigma = 23.51441 1.818897e-09 delta.llik = 0.000272052 
#&gt; Warning: solution lies close to zero for some positive variance components, their standard errors may not be valid</div><div class='input'><span class='no'>fit</span>$<span class='no'>df</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>fit</span>$<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fu'>pt</span>(<span class='no'>fit</span>$<span class='no'>s</span> / <span class='fu'>sqrt</span>(<span class='no'>fit</span>$<span class='no'>vs</span>), <span class='kw'>df</span> <span class='kw'>=</span> <span class='no'>fit</span>$<span class='no'>df</span>, <span class='kw'>lower.tail</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='no'>sets</span> <span class='kw'>&lt;-</span> <span class='fu'>list</span>(<span class='kw'>A</span> <span class='kw'>=</span> <span class='fu'>as.character</span>(<span class='fl'>1</span>:<span class='fl'>100</span>), <span class='kw'>B</span> <span class='kw'>=</span> <span class='fu'>as.character</span>(<span class='fl'>101</span>:<span class='fl'>1000</span>), <span class='kw'>C</span> <span class='kw'>=</span> <span class='fu'>as.character</span>(<span class='fl'>1001</span>:<span class='fl'>5000</span>), <span class='kw'>D</span> <span class='kw'>=</span> <span class='fu'>as.character</span>(<span class='fl'>5001</span>:<span class='fl'>10000</span>))

<span class='co'># Set test based on sums </span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>setTest</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>fit</span>$<span class='no'>s</span>**<span class='fl'>2</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"sum"</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>100</span>)

<span class='co'># Set test based on cvat </span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>setTest</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>fit</span>$<span class='no'>s</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='no'>W</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"cvat"</span>, <span class='kw'>nperm</span> <span class='kw'>=</span> <span class='fl'>100</span>)

<span class='co'># Set test based on hyperG </span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>setTest</span>(<span class='kw'>stat</span> <span class='kw'>=</span> <span class='no'>fit</span>$<span class='no'>p</span>, <span class='kw'>sets</span> <span class='kw'>=</span> <span class='no'>sets</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"hyperG"</span>, <span class='kw'>threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Peter Sørensen

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Peter Soerensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

